nixos/tests/nat.nix

```diff
@@ -1,100 +1,275 @@
-# This is a simple distributed test involving a topology with two
-# separate virtual networks - the "inside" and the "outside" - with a
-# client on the inside network, a server on the outside network, and a
-# router connected to both that performs Network Address Translation
-# for the client.
-import ./make-test-python.nix ({ pkgs, lib, withFirewall, nftables ? false, ... }:
+# This is a distributed test of the Network Address Translation involving a topology
+# with a router inbetween three separate virtual networks:
+# - "external" -- i.e. the internet,
+# - "internal" -- i.e. an office LAN,
+#
+# This test puts one server on each of those networks and its primary goal is to ensure that:
+# - server (named client in the code) in internal network can reach server (named server in the code) on the external network,
+# - server in external network can not reach server in internal network (skipped in some cases),
+# - when using externalIP, only the specified IP is used for NAT,
+# - port forwarding functionality behaves correctly
+#
+# The client is behind the nat (read: protected by the nat) and the server is on the external network, attempting to access services behind the NAT.
+
+import ./make-test-python.nix ({ pkgs, lib, withFirewall ? false, nftables ? false, ... }:
   let
     unit = if nftables then "nftables" else (if withFirewall then "firewall" else "nat");
+
+    routerAlternativeExternalIp = "192.168.2.234";
+
+    makeNginxConfig = hostname: {
+        enable = true;
+        virtualHosts."${hostname}" = {
+            root = "/etc";
+            locations."/".index = "hostname";
+            listen = [
+                {
+                    addr = hostname;
+                    port = 80;
+                }
+                {
+                    addr = hostname;
+                    port = 8080;
+                }
+            ];
+        };
+    };
+
+    makeCommonConfig = hostname: {
+      services.nginx = makeNginxConfig hostname;
+      services.vsftpd = {
+        enable = true;
+        anonymousUser = true;
+        localRoot = "/etc/";
+        extraConfig = ''
+          pasv_min_port=51000
+          pasv_max_port=51999
+        '';
+      };
+
+      # Disable eth0 autoconfiguration
+      networking.useDHCP = false;
+
+      environment.systemPackages = [
+        (pkgs.writeScriptBin "check-connection"
+          ''
+            #!/usr/bin/env bash
+
+            set -e
+
+            if [[ "$2" == "" || "$3" == "" || "$1" == "--help" || "$1" == "-h" ]];
+            then
+                echo "check-connection <target-address> <target-hostname> <[expect-success|expect-failure]>"
+                exit 1
+            fi
+
+            ADDRESS="$1"
+            HOSTNAME="$2"
+
+            function test_icmp() { timeout 3 ping -c 1 $ADDRESS; }
+            function test_http() { [[ `timeout 3 curl $ADDRESS` == "$HOSTNAME" ]] || echo "hello world" 1>&2; }
```

Comment:
remove debug print?

Reply:
Quite embarrassingly, this debug print actually hides a test failure. I've fixed this and will get this pushed ASAP. Thanks for finding this. (The tests were the issue, not the NAT)